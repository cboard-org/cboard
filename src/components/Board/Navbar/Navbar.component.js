import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import classNames from 'classnames';
import Tooltip from 'material-ui/Tooltip';
import IconButton from 'material-ui/IconButton';
import ArrowBackIcon from 'material-ui-icons/ArrowBack';
import LockOutlineIcon from 'material-ui-icons/LockOutline';
import LockOpenIcon from 'material-ui-icons/LockOpen';
import SettingsIcon from 'material-ui-icons/Settings';

import messages from './Navbar.messages';
import './Navbar.css';

Navbar.propTypes = {
  /**
   * @ignore
   */
  className: PropTypes.string,
  /**
   * @ignore
   */
  classes: PropTypes.object,
  /**
   * @ignore
   */
  intl: intlShape.isRequired,
  /**
   * Bar title
   */
  title: PropTypes.string,
  /**
   * If disabled, navigation is disabled
   */
  disabled: PropTypes.bool,
  /**
   * If enabled, navigation bar is locked Todo: shouldn't be here - mixing concerns
   */
  isLocked: PropTypes.bool,
  /**
   * Callback fired when clicking on back button
   */
  onBackClick: PropTypes.func,
  /**
   * Callback fired when clicking on lock button
   */
  onLockClick: PropTypes.func,
  /**
   * Callback fired when clicking on settings button
   */
  onSettingsClick: PropTypes.func
};

function Navbar({
  className,
  classes,
  intl,
  title,
  disabled,
  isLocked,
  onBackClick,
  onLockClick,
  onSettingsClick
}) {
  return (
    <div className={classNames('Navbar', className)}>
      <h2 className="Navbar__title">{title}</h2>
      <div className="Navbar__group Navbar__group--start">
        {!disabled && (
          <Tooltip title={intl.formatMessage(messages.back)} placement="bottom">
            <IconButton
              className="back-button"
              focusRipple={true}
              aria-label={intl.formatMessage(messages.back)}
              onClick={onBackClick}
              color="inherit"
            >
              <ArrowBackIcon />
            </IconButton>
          </Tooltip>
        )}

        {disabled && (
          <IconButton
            className="back-button"
            focusRipple={true}
            aria-label={intl.formatMessage(messages.back)}
            disabled={disabled}
            onClick={onBackClick}
            color="inherit"
            style={{
              color: 'rgba(255, 255, 255, 0.26)'
            }}
          >
            <ArrowBackIcon />
          </IconButton>
        )}
      </div>
      <div className="Navbar__group Navbar__group--end">
        {!isLocked && (
          <Tooltip
            title={intl.formatMessage(messages.settings)}
            placement="bottom"
          >
            <IconButton
              focusRipple={true}
              aria-label={intl.formatMessage(messages.settings)}
              color="inherit"
              onClick={onSettingsClick}
            >
              <SettingsIcon />
            </IconButton>
          </Tooltip>
        )}
        <Tooltip title={intl.formatMessage(messages.lock)} placement="bottom">
          <IconButton
            focusRipple={true}
            aria-label={intl.formatMessage(messages.lock)}
            color="inherit"
            onClick={onLockClick}
          >
            {isLocked ? <LockOutlineIcon /> : <LockOpenIcon />}
          </IconButton>
        </Tooltip>
      </div>
    </div>
  );
}

export default injectIntl(Navbar);
