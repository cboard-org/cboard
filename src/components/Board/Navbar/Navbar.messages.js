import { defineMessages } from 'react-intl';

export default defineMessages({
  back: {
    id: 'cboard.containers.Board.back',
    defaultMessage: 'Back'
  },
  lock: {
    id: 'cboard.containers.Board.lock',
    defaultMessage: 'Lock'
  },
  undo: {
    id: 'cboard.containers.Board.undo',
    defaultMessage: 'Undo'
  }
});
