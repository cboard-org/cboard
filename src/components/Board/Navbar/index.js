export { default } from './Navbar.component';
