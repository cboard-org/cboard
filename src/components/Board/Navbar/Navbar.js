import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import classNames from 'classnames';
import isMobile from 'ismobilejs';
import copy from 'copy-to-clipboard';
import { Scannable } from 'react-scannable';
import { IconButton } from '@material-ui/core';
import ScannerDeactivateIcon from '@material-ui/icons/ExploreOff';

import BoardShare from '../BoardShare';
import FullScreenButton from '../../UI/FullScreenButton';
import PrintBoardButton from '../../UI/PrintBoardButton';
import LockToggle from '../../UI/LockToggle';
import BackButton from '../../UI/BackButton';
import SettingsButton from '../../UI/SettingsButton';
import messages from '../Board.messages';

import './Navbar.css';
import { injectIntl } from 'react-intl';

class Navbar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      backButton: false,
      openShareDialog: false,
      deactivateScannerButton: false
    };
  }

  onShareClick = () => {
    let nativeShare = false;
    // https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share#Examples
    if (
      this.props.board &&
      this.props.board.isPublic &&
      window &&
      window.navigator &&
      window.navigator.share
    ) {
      try {
        window.navigator.share({
          title: this.props.board.name,
          text: this.props.board.name,
          url: window.location.href
        });
        nativeShare = true;
      } catch (e) {}
    }

    if (!nativeShare) {
      this.setState({ openShareDialog: true });
    }
  };

  onShareClose = () => {
    this.setState({ openShareDialog: false });
  };

  publishBoard = () => {
    this.props.publishBoard();
  };

  copyLinkAction = () => {
    copy(window.location.href);
    const copyMessage = this.props.intl.formatMessage(messages.copyMessage);

    this.props.showNotification(copyMessage);
  };

  onScannableFocus = property => () => {
    if (!this.state[property]) {
      this.setState({ [property]: true });
    }
  };

  onScannableBlur = property => () => {
    if (this.state[property]) {
      this.setState({ [property]: false });
    }
  };

  render() {
    const {
      className,
      intl,
      board,
      userData,
      title,
      disabled,
      isLocked,
      isScannerActive,
      onBackClick,
      onDeactivateScannerClick,
      onLockClick,
      onLockNotify
    } = this.props;

    const isPublic = board && board.isPublic;
    const isOwnBoard = board && board.email === userData.email;

    return (
      <div className={classNames('Navbar', className)}>
        {isLocked && <h2 className="Navbar__title">{title}</h2>}
        <div className="Navbar__group Navbar__group--start">
          <div className={this.state.backButton ? 'scanner__focused' : ''}>
            <Scannable
              disabled={disabled}
              onFocus={this.onScannableFocus('backButton')}
              onBlur={this.onScannableBlur('backButton')}
            >
              <BackButton disabled={disabled} onClick={onBackClick} />
            </Scannable>
          </div>
          {isScannerActive && (
            <div
              className={
                this.state.deactivateScannerButton ? 'scanner__focused' : ''
              }
            >
              <Scannable
                onFocus={this.onScannableFocus('deactivateScannerButton')}
                onBlur={this.onScannableBlur('deactivateScannerButton')}
              >
                <IconButton
                  className="Navbar__deactivateScanner"
                  onClick={onDeactivateScannerClick}
                >
                  <ScannerDeactivateIcon />
                </IconButton>
              </Scannable>
            </div>
          )}
        </div>
        <div className="Navbar__group Navbar__group--end">
          {!isLocked && (
            <React.Fragment>
              <PrintBoardButton />
              {!isMobile.any && <FullScreenButton />}
              <SettingsButton component={Link} to="/settings" />
              <BoardShare
                label={intl.formatMessage(messages.share)}
                intl={this.props.intl}
                isPublic={isPublic}
                isOwnBoard={isOwnBoard}
                onShareClick={this.onShareClick}
                onShareClose={this.onShareClose}
                publishBoard={this.publishBoard}
                copyLinkAction={this.copyLinkAction}
                open={this.state.openShareDialog}
                url={window.location.href}
              />
            </React.Fragment>
          )}

          <LockToggle
            locked={isLocked}
            onLockTick={onLockNotify}
            onClick={onLockClick}
          />
        </div>
      </div>
    );
  }
}

Navbar.propTypes = {
  /**
   * @ignore
   */
  className: PropTypes.string,
  /**
   * Bar title
   */
  title: PropTypes.string,
  /**
   * If disabled, navigation is disabled
   */
  disabled: PropTypes.bool,
  /**
   * If enabled, navigation bar is locked Todo: shouldn't be here - mixing concerns
   */
  isLocked: PropTypes.bool,
  /**
   * Callback fired when clicking on back button
   */
  onBackClick: PropTypes.func,
  /**
   * Callback fired when clicking on lock button
   */
  onLockClick: PropTypes.func,
  isScannerActive: PropTypes.bool,
  onDeactivateScannerClick: PropTypes.func
};

export default injectIntl(Navbar);
