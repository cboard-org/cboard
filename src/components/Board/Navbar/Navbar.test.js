import React from 'react';
import { shallow, mount } from 'enzyme';

import Navbar from './Navbar';

const mockBoard = {
  name: 'tewt',
  id: '12345678901234567',
  tiles: [{ id: '1234', loadBoard: '456456456456456456456' }],
  isPublic: false,
  email: 'asd@qwe.com',
  markToUpdate: true
};

describe('NavBar tests', () => {
  const COMPONENT_PROPS = {
    intl: {},
    board: mockBoard,
    userData: { email: 'qa@qa.com' },
    onLockNotify: jest.fn(),
    className: 'string',
    title: 'string',
    disabled: false,
    isLocked: false,
    onBackClick: jest.fn(),
    onLockClick: jest.fn(),
    isScannerActive: false,
    onDeactivateScannerClick: jest.fn()
  };

  it('renders without crashing', () => {
    const wrapper = shallow(<Navbar {...COMPONENT_PROPS} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders without crashing locked', () => {
    const props = {
      ...COMPONENT_PROPS,
      isLocked: true
    };
    const wrapper = shallow(<Navbar {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('check share', () => {
    const wrapper = shallow(<Navbar {...COMPONENT_PROPS} />);
    wrapper.prop('onLockClick')();
  });
});
