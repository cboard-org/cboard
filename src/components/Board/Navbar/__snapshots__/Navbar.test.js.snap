// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar tests renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    board={
      Object {
        "email": "asd@qwe.com",
        "id": "12345678901234567",
        "isPublic": false,
        "markToUpdate": true,
        "name": "tewt",
        "tiles": Array [
          Object {
            "id": "1234",
            "loadBoard": "456456456456456456456",
          },
        ],
      }
    }
    className="string"
    disabled={false}
    intl={Object {}}
    isLocked={false}
    isScannerActive={false}
    onBackClick={[MockFunction]}
    onDeactivateScannerClick={[MockFunction]}
    onLockClick={[MockFunction]}
    onLockNotify={[MockFunction]}
    title="string"
    userData={
      Object {
        "email": "qa@qa.com",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "board": Object {
        "email": "asd@qwe.com",
        "id": "12345678901234567",
        "isPublic": false,
        "markToUpdate": true,
        "name": "tewt",
        "tiles": Array [
          Object {
            "id": "1234",
            "loadBoard": "456456456456456456456",
          },
        ],
      },
      "className": "string",
      "disabled": false,
      "intl": Object {
        "formatMessage": [Function],
      },
      "isLocked": false,
      "isScannerActive": false,
      "onBackClick": [MockFunction],
      "onDeactivateScannerClick": [MockFunction],
      "onLockClick": [MockFunction],
      "onLockNotify": [MockFunction],
      "title": "string",
      "userData": Object {
        "email": "qa@qa.com",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "board": Object {
          "email": "asd@qwe.com",
          "id": "12345678901234567",
          "isPublic": false,
          "markToUpdate": true,
          "name": "tewt",
          "tiles": Array [
            Object {
              "id": "1234",
              "loadBoard": "456456456456456456456",
            },
          ],
        },
        "className": "string",
        "disabled": false,
        "intl": Object {
          "formatMessage": [Function],
        },
        "isLocked": false,
        "isScannerActive": false,
        "onBackClick": [MockFunction],
        "onDeactivateScannerClick": [MockFunction],
        "onLockClick": [MockFunction],
        "onLockNotify": [MockFunction],
        "title": "string",
        "userData": Object {
          "email": "qa@qa.com",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`NavBar tests renders without crashing locked 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navbar
    board={
      Object {
        "email": "asd@qwe.com",
        "id": "12345678901234567",
        "isPublic": false,
        "markToUpdate": true,
        "name": "tewt",
        "tiles": Array [
          Object {
            "id": "1234",
            "loadBoard": "456456456456456456456",
          },
        ],
      }
    }
    className="string"
    disabled={false}
    intl={Object {}}
    isLocked={true}
    isScannerActive={false}
    onBackClick={[MockFunction]}
    onDeactivateScannerClick={[MockFunction]}
    onLockClick={[MockFunction]}
    onLockNotify={[MockFunction]}
    title="string"
    userData={
      Object {
        "email": "qa@qa.com",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "board": Object {
        "email": "asd@qwe.com",
        "id": "12345678901234567",
        "isPublic": false,
        "markToUpdate": true,
        "name": "tewt",
        "tiles": Array [
          Object {
            "id": "1234",
            "loadBoard": "456456456456456456456",
          },
        ],
      },
      "className": "string",
      "disabled": false,
      "intl": Object {
        "formatMessage": [Function],
      },
      "isLocked": true,
      "isScannerActive": false,
      "onBackClick": [MockFunction],
      "onDeactivateScannerClick": [MockFunction],
      "onLockClick": [MockFunction],
      "onLockNotify": [MockFunction],
      "title": "string",
      "userData": Object {
        "email": "qa@qa.com",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "board": Object {
          "email": "asd@qwe.com",
          "id": "12345678901234567",
          "isPublic": false,
          "markToUpdate": true,
          "name": "tewt",
          "tiles": Array [
            Object {
              "id": "1234",
              "loadBoard": "456456456456456456456",
            },
          ],
        },
        "className": "string",
        "disabled": false,
        "intl": Object {
          "formatMessage": [Function],
        },
        "isLocked": true,
        "isScannerActive": false,
        "onBackClick": [MockFunction],
        "onDeactivateScannerClick": [MockFunction],
        "onLockClick": [MockFunction],
        "onLockNotify": [MockFunction],
        "title": "string",
        "userData": Object {
          "email": "qa@qa.com",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
